File: api-dotnet/Controllers/RegexController.cs
using Microsoft.AspNetCore.Mvc;
using RegExTester.Api.DotNet.Models;
using RegExTester.Api.DotNet.Services;
using System;
using System.Diagnostics;
using System.Threading;

namespace RegExTester.Api.DotNet.Controllers
{
    /// <summary>Executes .NET regular expressions against supplied input text.</summary>
    [ApiController]
    [Route("api/regex")]
    [Produces("application/json")]
    public class RegExController : Controller
    {
        public IRegExProcessor RegExProcessor { get; set; }
        public ITelemetryService TelemetryService { get; set; }

        public RegExController(IRegExProcessor regExProcessor, ITelemetryService telemetryService)
        {
            this.RegExProcessor = regExProcessor;
            this.TelemetryService = telemetryService;
        }

        /// <summary>Run a regular expression and return all matches.</summary>
        /// <remarks>
        /// Applies the provided regex pattern to the input text using the specified option flags.
        /// All string fields are Base64Url-encoded by the frontend before submission but
        /// the API itself accepts plain UTF-8 JSON strings.
        ///
        /// **Timeout:** The regex engine enforces a 15-second match timeout; the HTTP request
        /// has a 5-second middleware timeout. If either is exceeded an error message is returned
        /// in the `error` field rather than throwing an HTTP error.
        /// </remarks>
        /// <param name="model">Pattern, text, optional replacement string, and regex option flags.</param>
        /// <param name="cancellationToken">Request cancellation token.</param>
        /// <returns>Match results including groups and (optionally) captures.</returns>
        /// <response code="200">Regex executed successfully; inspect <c>error</c> field for pattern errors.</response>
        /// <response code="400">Request body failed model validation (e.g. pattern &gt; 512 chars).</response>
        [HttpPost]
        [ProducesResponseType(typeof(RegexResult), 200)]
        [ProducesResponseType(400)]
        public ActionResult Post([FromBody] Input model, CancellationToken cancellationToken)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var stopwatch = Stopwatch.StartNew();
            var result = RegExProcessor.Matches(model.Pattern, model.Text, model.Replace, model.Options);
            stopwatch.Stop();

            // Fire-and-forget: TelemetryService dispatches the Cosmos write on a background task
            // and never throws, so this call never delays or affects the response below.
            TelemetryService.RecordTelemetry(Request, model, stopwatch.Elapsed, result.Matches.Count, result.Error);

            return Json(result);
        }
    }
}

File: api-nodejs/src/controllers/regexController.js
import { RegexProcessor } from '../services/regexProcessor.js';
import { telemetryService } from '../services/telemetryService.js';

export const regexController = {
  /**
   * @openapi
   * /api/regex:
   *   post:
   *     tags: [RegEx]
   *     summary: Run a regular expression and return all matches.
   *     description: |
   *       Applies the provided regex pattern to the input text using the specified option flags.
   *       All string fields are Base64Url-encoded by the Vue.js frontend before submission but
   *       the API itself accepts plain UTF-8 JSON strings.
   *
   *       **Timeout:** The regex engine enforces a 15-second match timeout; the HTTP request
   *       has a 5-second middleware timeout. If either is exceeded an error message is returned
   *       in the `error` field rather than throwing an HTTP error.
   *
   *       **Request size:** The raw request body must not exceed `limits.maxRequestBodyBytes`
   *       (see `GET /api/capabilities`) — currently 8192 bytes. This comfortably fits the
   *       largest valid payload (`pattern` + `text` + `replace` at their documented maximum
   *       lengths, plus JSON overhead and multi-byte UTF-8) while still bounding request size.
   *     requestBody:
   *       description: Regex evaluation request.
   *       required: true
   *       content:
   *         application/json:
   *           schema:
   *             $ref: '#/components/schemas/Input'
   *     responses:
   *       200:
   *         description: Regex executed successfully; inspect `error` field for pattern errors.
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/RegexResult'
   *       400:
   *         description: |
   *           Request body failed model validation (e.g. `pattern` > 512 chars), or the request
   *           body is not valid JSON.
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/ProblemDetails'
   *       413:
   *         description: |
   *           The raw request body exceeded `limits.maxRequestBodyBytes` (8192 bytes).
   *         content:
   *           application/json:
   *             schema:
   *               $ref: '#/components/schemas/ProblemDetails'
   */
  match(req, res) {
    const { pattern, text, replace, options } = req.body;

    // Validation
    const errors = {};
    if (pattern != null && pattern.length > 512) {
      errors.pattern = ['Pattern must be 512 characters or fewer.'];
    }
    if (text != null && text.length > 1024) {
      errors.text = ['Text must be 1024 characters or fewer.'];
    }
    if (replace != null && replace.length > 1024) {
      errors.replace = ['Replace must be 1024 characters or fewer.'];
    }

    if (Object.keys(errors).length > 0) {
      return res.status(400).json({
        type: 'https://tools.ietf.org/html/rfc9110#section-15.5.1',
        title: 'One or more validation errors occurred.',
        status: 400,
        errors,
      });
    }

    const start = process.hrtime.bigint();
    const result = RegexProcessor.match(pattern, text, replace, options ?? 0);
    const durationMs = Number(process.hrtime.bigint() - start) / 1e6;
    telemetryService.sendTelemetry(req, { pattern, text, replace, options }, {
      durationMs: Math.round(durationMs),
      matchCount: result.matches.length,
      error: result.error,
    });
    res.json(result);
  },
};
